languages/elixir/exercises/concept/recursion/.docs/after.md

```diff
@@ -0,0 +1,72 @@
+Recursive functions are functions that call themselves.
+
+A recursive function needs to have at least one _base case_ and at least one _recursive case_.
+
+A _base case_ returns a value without calling the function again. A _recursive case_ calls the function again, modifying the input so that it will at some point match the base case.
+
+Very often, each case is written in its own function clause.
+
+```elixir
+# base case
+def count([]), do: 0
+
+# recursive case
+def count([_head | tail]), do: 1 + count(tail)
+```
+
+A recursive function can have many base cases and/or many recursive cases. For example [the fibonacci sequence][fibonacci] is a recursive sequence with two base cases:
+
+```elixir
+def fibonacci(0), do: 0
+def fibonacci(1), do: 1
+def fibonacci(n), do: fibonacci(n - 1) + fibonacci(n - 2)
+```
+
+Counting the number of occurrences of x in a list has two recursive cases:
+
+```elixir
+def count_occurrences([], _x), do: 0
+def count_occurrences([x | tail], x), do: 1 + count_occurrences(tail, x)
+def count_occurrences([_ | tail], x), do: count_occurrences(tail, x)
+```
+
+## Loops through recursion
+
+Due to immutability, loops in Elixir are written differently from imperative languages. For example, in C, a loop would look like this:
+
+```c
+for(i = 0; i < sizeof(array); i++) {
+  printf("%d\n", array[i]);
+}
+```
+
+In a functional language, mutating `i` (by calling `i++`) is not possible. Thus, loops have to be implemented with recursion.
+
+The equivalent of a `for` loop in Elixir would look like this:
```

Comment:
I'm not sure if this C example is actually helping or possibly confusing. One would have to know about quite a bit of C syntax (for, mutation, sizeof, printf) to be able to make sense of it. Personally, I think I'd prefer _not_ having the C example, but I do understand why you added the example. @ihid what do you think?

Reply:
I like the example a lot. It made it immediately clear to me. But being in `c` is offputting I think. I'd suggest some sort of pseudocode. For example

```
Due to immutability, loops in Elixir are written differently from imperative languages. For example, loops commonly look like:
```
```
for(i = 0; i < array.size; i++) {
  // Do something with array[i]
}
```

Reply:
I like pseudocode!

Reply:
Excellent idea! Now I don't have to worry if I wrote that C code correctly, it's been years 😁